crates/bindings-csharp/BSATN.Codegen/Utils.cs

```diff
@@ -126,25 +113,19 @@ static string GetTypeInfoForNamedType(INamedTypeSymbol type)
                 }
                 return $"SpacetimeDB.BSATN.Enum<{SymbolToName(type)}>";
             }
-            var result = type.OriginalDefinition.ToString() switch
+            var fullName = type.ToDisplayString(SymbolFormatForBSATN);
+            string result;
+            if (fullName.StartsWith("System."))
             {
-                // {U/I}{128/256} are not treated by C# as regular primitives, so we need to match them by type name.
-                "System.Int128" => "SpacetimeDB.BSATN.I128",
-                "System.UInt128" => "SpacetimeDB.BSATN.U128",
-                "SpacetimeDB.I128" => "SpacetimeDB.BSATN.I128Stdb",
-                "SpacetimeDB.U128" => "SpacetimeDB.BSATN.U128Stdb",
-                "SpacetimeDB.I256" => "SpacetimeDB.BSATN.I256",
-                "SpacetimeDB.U256" => "SpacetimeDB.BSATN.U256",
-                "System.Collections.Generic.List<T>" => $"SpacetimeDB.BSATN.List",
-                "System.Collections.Generic.Dictionary<TKey, TValue>" =>
-                    $"SpacetimeDB.BSATN.Dictionary",
-                // If we're here, then this is nullable *value* type like `int?`.
-                "System.Nullable<T>" => $"SpacetimeDB.BSATN.ValueOption",
-                var name when name.StartsWith("System.") => throw new InvalidOperationException(
-                    $"Unsupported system type {name}"
-                ),
-                _ => $"{SymbolToName(type)}.BSATN",
-            };
+                // C# doesn't have traits, so we put all the implementations for built-in types
+                // in a special namespace with 1:1 name mapping.
+                result = $"SpacetimeDB.BSATN.SystemImpls.{fullName[7..]}";
```

Comment:
There's a whole lot of System.* types, do we really want to allow System.Reflection.Metadata.MetadataReader to be caught by this for example?

Reply:
IMO yes, one way or another we should (and will) get a compilation error for unsupported types.